tests/PolyType.Tests/TypeShapeProviderStaticTests.cs

```diff
@@ -0,0 +1,146 @@
+﻿using System.Reflection;
+
+namespace PolyType.Tests;
+
+public class TypeShapeProviderStaticTests
+{
+    [Fact]
+    public void Combine_EmptyList()
+    {
+        ITypeShapeProvider aggregate = TypeShapeProvider.Combine();
+        Assert.NotNull(aggregate);
+        Assert.Null(aggregate.GetShape(typeof(int)));
+    }
+
+    [Fact]
+    public void Combine_One_Null()
+    {
+        ITypeShapeProvider aggregate = TypeShapeProvider.Combine(new MockTypeShapeProvider());
+        Assert.Null(aggregate.GetShape(typeof(int)));
+    }
+
+
+    [Fact]
+    public void Combine_One_NonNull()
+    {
+        MockTypeShapeProvider first = new()
+        {
+            Shapes =
+            {
+                [typeof(int)] = new MockTypeShape<int>(),
+            },
+        };
+        ITypeShapeProvider aggregate = TypeShapeProvider.Combine(first);
+        Assert.Same(first.Shapes[typeof(int)], aggregate.GetShape(typeof(int)));
+    }
+
+    [Fact]
+    public void Combine_Many()
+    {
+        MockTypeShapeProvider first = new()
+        {
+            Shapes =
+            {
+                [typeof(int)] = new MockTypeShape<int>(),
+            },
+        };
+        MockTypeShapeProvider second = new()
+        {
+            Shapes =
+            {
+                [typeof(string)] = new MockTypeShape<int>(),
+            },
+        };
+        ITypeShapeProvider aggregate = TypeShapeProvider.Combine(first, second);
+        Assert.Same(first.Shapes[typeof(int)], aggregate.GetShape(typeof(int)));
+        Assert.Same(second.Shapes[typeof(string)], aggregate.GetShape(typeof(string)));
+        Assert.Null(aggregate.GetShape(typeof(bool)));
+    }
+
+    /// <summary>
+    /// Validates that the aggregating provider does not cache results.
```

Comment:
While I agree that intermediate providers shouldn't be caching results (caching should be done at level of the consuming visitors), I think it's probably fine to leave as undefined behavior.

Reply:
If we want to document it as undefined, allowing us the freedom to add a cache later, that's... potentially breaking anyway as folks may take dependencies on the old behavior, but I'd go along with it.
But I'd still add tests, so that if the behavior changes, the tests call it out, and we can second guess ourselves and update the tests if necessary, since by that point, it _will_ be an intentional change and having tests that only need a couple changes will help set us up for later too.

Reply:
Sounds good.